sklearn/cluster/_k_means_elkan.pyx

```diff
@@ -0,0 +1,215 @@
+        x = X + sample * n_features
+        d_c = euclidian_dist(x, centers, n_features)
+        lower_bounds[sample, 0] = d_c
+        for j in range(1, n_clusters):
+            if d_c > center_distances[c_x, j]:
+                c = centers + j * n_features
+                dist = euclidian_dist(x, c, n_features)
+                lower_bounds[sample, j] = dist
+                if dist < d_c:
+                    d_c = dist
+                    c_x = j
+        labels[sample] = c_x
+        upper_bounds[sample] = d_c
+
+
+def k_means_elkan(np.ndarray[np.float64_t, ndim=2, mode='c'] X_, int n_clusters,
+                  np.ndarray[np.float64_t, ndim=2, mode='c'] init,
+                  float tol=1e-4, int max_iter=30, verbose=False):
+    """Run Elkan's k-means.
+
+    Parameters
+    ----------
+    X_ : nd-array, shape (n_samples, n_features)
+
+    n_clusters : int
+        Number of clusters to find.
+
+    init : nd-array, shape (n_clusters, n_features)
+        Initial position of centers.
+
+    tol : float, default=1e-4
+        The relative increment in cluster means before declaring convergence.
+
+    max_iter : int, default=30
+        Maximum number of passes over the dataset.
+
+    verbose : bool, default=False
+        Whether to be verbose.
+
+    """
+    #initialize
+    cdef np.ndarray[np.float64_t, ndim=2, mode='c'] centers_ = init
+    cdef double* centers_p = <double*>centers_.data
+    cdef double* X_p = <double*>X_.data
+    cdef double* x_p
+    cdef Py_ssize_t n_samples = X_.shape[0]
+    cdef Py_ssize_t n_features = X_.shape[1]
+    cdef int point_index, center_index, label
+    cdef float upper_bound, distance
+    cdef double[:, :] center_distances = euclidean_distances(centers_) / 2.
+    cdef double[:, :] lower_bounds = np.zeros((n_samples, n_clusters))
+    cdef double[:] distance_next_center
+    labels_ = np.empty(n_samples, dtype=np.int32)
+    cdef int[:] labels = labels_
+    upper_bounds_ = np.empty(n_samples, dtype=np.float)
+    cdef double[:] upper_bounds = upper_bounds_
+
+    # Get the inital set of upper bounds and lower bounds for each sample.
+    update_labels_distances_inplace(X_p, centers_p, center_distances, labels,
+                                    lower_bounds, upper_bounds, n_samples,
+                                    n_features, n_clusters)
+    cdef np.uint8_t[:] bounds_tight = np.ones(n_samples, dtype=np.uint8)
+    cdef np.uint8_t[:] points_to_update = np.zeros(n_samples, dtype=np.uint8)
+    cdef np.ndarray[np.float64_t, ndim=2, mode='c'] new_centers
+
+    if max_iter <= 0:
+        raise ValueError('Number of iterations should be a positive number'
+        ', got %d instead' % max_iter)
+
+    for iteration in range(max_iter):
+        if verbose:
+            print("start iteration")
+
+        # we could find the closest center in O(n) but
+        # this does not seem to be the bottleneck
+        distance_next_center = np.sort(center_distances, axis=0)[1]
```

Comment:
Can you check if we gain anything by using `np.argpartition` here? there should be some gains for the n_clusters=1000 case at least.